app/controllers/incomeSources/cease/IncomeSourceEndDateController.scala

```diff
@@ -262,33 +263,43 @@ class IncomeSourceEndDateController @Inject()(val authenticate: AuthenticationPr
+                    s"$errorMessage")
+                  Future.failed(new Exception(s"$errorMessage"))
+                case (SelfEmployment, Some(incomeSourceId)) =>
+                  val result = Redirect(redirectAction)
+                  sessionService.setMongoKey(
+                    CeaseIncomeSourceData.dateCeasedField, validatedInput.date.toString, JourneyType(Cease, incomeSourceType)
+                  ).flatMap {
+                    case Right(true) => {
+                      sessionService.setMongoKey(
+                        CeaseIncomeSourceData.incomeSourceIdField, incomeSourceId, JourneyType(Cease, incomeSourceType)
+                      ).flatMap {
+                        case Right(true) =>
+                          Future.successful(result)
+                        case _ => Future.failed(new Error(s"Failed to set income source id in session storage. incomeSourceType: $incomeSourceType. incomeSourceType: $incomeSourceType"))
+                      }
+                    }
+                    case _ => Future.failed(new Error(s"Failed to set end date value in session storage. incomeSourceType: $incomeSourceType, incomeSourceType: $incomeSourceType"))
+                  }
 
-          case (SelfEmployment, None) =>
-            val errorMessage: String = s"[IncomeSourceEndDateController][handleSubmitRequest]: missing income source ID - $id."
-            Logger("application").error(s"${if (isAgent) "[Agent]"}" +
-              s"$errorMessage")
-            Future.failed(new Exception(s"$errorMessage"))
-
-          case (SelfEmployment, Some(incomeSourceId)) =>
-            val result = Redirect(redirectAction)
-            sessionService.set(result,
-              incomeSourceType.endDateSessionKey -> validatedInput.date.toString,
-              ceaseBusinessIncomeSourceId -> incomeSourceId
-            ).flatMap {
-              case Right(result) => Future.successful(result)
-              case Left(exception) => Future.failed(exception)
-            }
-
-          case _ =>
-            val session = incomeSourceType.endDateSessionKey -> validatedInput.date.toString
-            val result = Redirect(redirectAction)
-            sessionService.set(result, session).flatMap {
-              case Right(result) => Future.successful(result)
-              case Left(exception) => Future.failed(exception)
-            }
-
-        })
+                case _ =>
+                  val propertyEndDate = validatedInput.date.toString
+                  val result = Redirect(redirectAction)
+                  sessionService.setMongoKey(key = CeaseIncomeSourceData.dateCeasedField, value = propertyEndDate, journeyType = JourneyType(Cease, incomeSourceType)).flatMap {
+                    case Right(_) => Future.successful(result)
+                    case Left(exception) => Future.failed(exception)
+                  }
+              }
+            case false => Future.failed(new Error("Failed to create mongo session"))
```

Comment:
this is not reachable branch of the logic; result will be always "true" during evaluation;
you can simply flatMap with mo pattern matching for result value;